### PR TITLE
[EMB-324] Better analytics

### PR DESCRIPTION
**Pull request by aaxelb**
## Purpose
Implement an easier way to track user actions, using `data-` attributes

## Summary of Changes
- Support tracking click events by adding `data-` attrs (documented in the handbook and below)
- Add click tracking to registries pages
- Add page tracking to registries pages
- Add "dev options" menu, accessible by clicking on the red dev-mode banner in the lower left
  - "Display toast on tracked events" will display a toast on tracked events, for convenient manual testing of analytics
- Add custom `click` helper in `ember-osf-web/tests/helpers` that will error if the click target doesn't have a `data-analytics-name` attr
- Remove the `ember_registries_new_style` flag -- instead base new styles on the overview page's waffle flag. The `new_style` flag has never been used.

## Side Effects
- Adding `data-analytics-name` to reusable components may result in duplicate events if they're also tracked the "old" way.
- Because the analytics service is now used in an instance initializer, the singleton is instantiated before tests begin to run. Therefore, trying to stub it with `this.owner.register('service:analytics', MyStub)` has no effect. Instead, consider using [sinon](https://sinonjs.org/releases/v7.2.2/stubs/) to stub methods in place:
  ```ts
  import sinon from 'sinon';
  // ...
  const analytics = this.owner.lookup('service:analytics');
  const clickStub = sinon.stub(analytics, 'click');
  ```

## QA Notes
Should be no visible changes, except to the dev-mode banner.

## Ticket

https://openscience.atlassian.net/browse/EMB-324

# Reviewer Checklist

- [ ] meets requirements
- [ ] easy to understand
- [ ] DRY
- [ ] testable and includes test(s)
- [ ] changes described in `CHANGELOG.md`

-----
# Analytics docs (copied from handbook)
- Every meaningful user action should be tracked
- Adding tracking should be easy
- Forgetting to add tracking should be hard

## Tracking clicks automatically
Tracking clicks is as easy as adding a few `data-` attributes.

### `data-analytics-name`
If an element is named with a `data-analytics-name` attribute, click events on that element will be tracked:
```hbs
<button
    data-analytics-name='Create new project'
    onclick={{action ... }}
></button>
```

### `data-analytics-scope`
Use named scopes to provide more context with the event, like on which page or region of the page the event took place.
For example, a click on the button below will be tracked with the label "Dashboard - Control bar - Create new project":
```hbs
<div data-analytics-scope='Dashboard'>
    <div data-analytics-scope='Control bar'>
        <button data-analytics-name='Create new project'></button>
    </div>
</div>
```

### `data-analytics-category`
The event category generally refers to the type of thing being acted upon. For normal links (`<a>` or `role='link'`)
and buttons (`<button>` or `role='button'`) the category will be inferred as `link` and `button`, respectively.

If you want to categorize events another way (e.g. `file` for all events that act on file objects), you can add
`data-analytics-category='mycategory'` to the same element with `data-analytics-name`:
```hbs
<button
    data-analytics-name='Delete file'
    data-analytics-category='file'
    onclick={{action this.deleteFile file}}
>Delete file</button>
```

### `data-analytics-extra`
You can optionally fill in the `extra` field on the tracked event by setting `data-analytics-extra='extra data'`:
```hbs
<button
    data-analytics-name='Delete file'
    data-analytics-category='file'
    data-analytics-extra={{concat 'File name: ' file.name}}
    onclick={{action this.deleteFile file}}
>Delete file</button>
```

## Tracking events by hand
<aside>
    Note: It is currently easy to forget to track non-click events.
    Until we figure out a more automatic approach, try not to forget.
</aside>

For non-click events, or when the "automatic" approach doesn't fit, call `track()` on the `analytics` service:

```ts
@action
onKeyUp() {
    this.analytics.track('search', 'keyUp', 'Search box - Key up');
    // ...
}
```

Params for `track()`:
- `category`: string
- `actionName`: string
- `label`: string
- `extraInfo` (optional): string

## Tracking page views
<aside>
    Note: It is currently easy to forget to add page tracking.
    Until we figure out a more automatic approach, try not to forget.
</aside>

Call `trackPage()` on the `analytics` service in the `didTransition` hook for your route or a parent route:
```ts
// route.ts
@action
didTransition() {
    this.analytics.trackPage();
}
```

If the page is under a GUID route, pass the following two arguments to `trackPage`:
- `pagePublic`: boolean indicating whether the GUID referent is public or private
- `resourceType`: string for the API type of the GUID referent (e.g. `nodes`, `registrations`, etc.)

## Testing analytics
Any action worth testing is worth tracking, and vice versa.

### Custom `click` helper
The custom `click` helper will error if its target does not have a `data-analytics-name` attribute.
Use it in all your tests for pages/components that use the "automatic" analytics approach.

```ts
import { click } from 'ember-osf-web/tests/helpers';
// ...
await click('button');
```

### Manual testing
In development builds, events that would be tracked are logged to the console.

In addition, you can click on the "development mode" banner in the lower left to open dev options.
Check "Display toast for tracked events" to display a toast for tracked events.

### CODE REVIEW

**Comment by brianjgeiger** [comment]
screen shot 2019-01-09 at 1 41 06 pm

Well, _that's_ ominous.

File: app/services/analytics.ts [hunk 1]
```diff
@@ -1,19 +1,137 @@
...
+            if (role === 'button' || element.tagName === 'BUTTON') {
+                this.category = 'button';
+            } else if (role === 'link' || element.tagName === 'A') {
+                this.category = 'link';
+            } else if (element.tagName === 'INPUT' && element.hasAttribute('type')) {
+                this.category = element.getAttribute('type')!;
+            }
+        }
+
+        assert('Event category could not be inferred. It must be set explicitly.', Boolean(this.category));
```
Comment (brianjgeiger) [inline app/services/analytics.ts]: Drive-by: We do also have `role=tab` in the codebase. I'd check for that first unless you want it to be a button or a link (whichever it happened to be implemented as at the time)

File: lib/osf-components/addon/components/osf-navbar/template.hbs [hunk 2]
```diff
@@ -3,21 +3,24 @@
         <div class='container'>
             <div class='navbar-header'>
                 {{! template-lint-disable no-implicit-this }}
-                {{#global-link-to 
-                    'home' 
-                    data-test-nav-home-link 
-                    (html-attributes aria-label=(t 'navbar.go_home')) 
+                {{#global-link-to
```
Comment (brianjgeiger) [inline lib/osf-components/addon/components/osf-navbar/template.hbs]: More drive-by: this should be a Link

File: lib/registries/addon/components/registries-navbar/template.hbs [hunk 3]
```diff
@@ -1,105 +1,192 @@
...
-            {{#global-link-to data-test-brand 'home' (html-attributes aria-label=(t 'navbar.go_home'))}}
-                <div local-class="Logo"></div>
-            {{/global-link-to}}
-
-            {{#bs-dropdown as |dd|}}
-                {{#dd.toggle data-test-service classNames=(local-class 'Service Dropdown')}}
-                    <span local-class="HideOnMobile">{{t 'general.OSF'}}</span>
+            <Link
+                data-analytics-name='Brand'
+                data-test-brand
```
Comment (brianjgeiger) [inline lib/registries/addon/components/registries-navbar/template.hbs]: We've been putting the data-test as the first item in the component partially because it had to be for curly-brace components but mostly because it makes for easy identification of the component.

File: mirage/scenarios/default.ts [hunk 4]
```diff
@@ -17,14 +17,7 @@ const {
...
 export default function(server: Server) {
-    let currentUser = null;
-
-    if (defaultLoggedOut) {
-        currentUser = server.create('user');
-        server.create('root', { currentUser: null });
-    } else {
-        currentUser = server.create('user', 'loggedIn');
-    }
+    const currentUser = defaultLoggedOut ? server.create('user') : server.create('user', 'loggedIn');
```
Comment (brianjgeiger) [inline mirage/scenarios/default.ts]: `currentUser: null` is important unless you fixed the pages that have a problem with not finding a user. Can I remember which pages? I can not.

Comment (aaxelb) [reply mirage/scenarios/default.ts]: I didn't fix those pages, but it's not necessary to create a root here at all. If there's no user with the `loggedIn` trait, the [root created by default has `currentUser: null`](https://github.com/CenterForOpenScience/ember-osf-web/blob/ca61f3448da94e039ea2d6126a16003d8f22452a/mirage/views/root.ts#L4).

File: app/services/analytics.ts (hunk 1 above)
Comment (aaxelb) [reply app/services/analytics.ts]: Ah, that makes sense. If a `role` is explicitly set, should we always just use that for the category (assuming no `data-analytics-category`)?

File: lib/registries/addon/components/registries-navbar/template.hbs (hunk 3 above)
Comment (aaxelb) [reply lib/registries/addon/components/registries-navbar/template.hbs]: I'll change that to be consistent, but it does feel like `data-test-` and `data-analytics-name` are now mildly redundant...

File: app/services/analytics.ts (hunk 1 above)
Comment (aaxelb) [reply app/services/analytics.ts]: I [switched](https://github.com/CenterForOpenScience/ember-osf-web/pull/548/commits/7eabacd114e8253af91532fe412d975a9cc8bdc3) to using any `role` as category... does that seem reasonable?

File: mirage/scenarios/default.ts (hunk 4 above)
Comment (brianjgeiger) [reply mirage/scenarios/default.ts]: Very good. Late November is probably after the last time I saw the problem 👍

File: app/services/analytics.ts (hunk 1 above)
Comment (brianjgeiger) [reply app/services/analytics.ts]: That does seem reasonable.

File: app/services/analytics.ts [hunk 5]
```diff
@@ -1,19 +1,136 @@
...
+} = config;
+
+interface TrackedData {
+    category?: string;
+    action?: string;
+    extra?: string;
+    label: string;
+}
+
+function logEvent(analytics: Analytics, title: string, data: any) {
```
Comment (jamescdavis) [inline app/services/analytics.ts]: Should be `data: object`?

File: lib/registries/addon/components/registries-navbar/template.hbs (hunk 3 above)
Comment (brianjgeiger) [reply lib/registries/addon/components/registries-navbar/template.hbs]: Yeah, I don't see any reason not to get rid of data-test on elements that have data-analytics if you change the tests and inform QA about the change so they can also update.

File: tests/integration/analytics-test.ts [hunk 6]
```diff
@@ -0,0 +1,228 @@
+import { click, render } from '@ember/test-helpers';
+import { setupRenderingTest } from 'ember-qunit';
+import { TestContext } from 'ember-test-helpers';
+import hbs from 'htmlbars-inline-precompile';
+import { module, test } from 'qunit';
+import sinon, { SinonStub } from 'sinon';
+
+// import { initialize } from 'ember-osf-web/instance-initializers/analytics';
```
Comment (jamescdavis) [inline tests/integration/analytics-test.ts]: pls remove

**Comment by aaxelb** [comment]
@jamescdavis done!

File: lib/registries/addon/components/side-nav/x-link/template.hbs [hunk 7]
```diff
@@ -7,7 +7,7 @@
         ...attributes
     >
         <span local-class='Label'>
-            <FaIcon @icon={{@icon}} @fixedWidth={{true}} @class={{local-class 'Icon'}} />
+            <FaIcon @name={{@icon}} @fixedWidth={{true}} @class={{local-class 'Icon'}} />
```
Comment (jamescdavis) [inline lib/registries/addon/components/side-nav/x-link/template.hbs]: I'm pretty sure what you had was correct: https://github.com/martndemus/ember-font-awesome/blob/v3.1.1/addon/components/fa-icon.js#L51-L54

Comment (aaxelb) [reply lib/registries/addon/components/side-nav/x-link/template.hbs]: whoops, I switched back and forth and dropped this one along the way. fixed!
